design-proposals/export-import/vm-export-import.md

```diff
@@ -0,0 +1,166 @@
+
+### User Flow
+
+1.  Create a VirtualMachineExport resource
+2.  Wait for the `Ready` condition of the VirtualMachineExport to be `True`
+3.  Download the Virtual Machine Archive by getting the URL in `status.url`
+
+### Export a VirtualMachine
+
+``` yaml
+apiVersion: export.kubevirt.io/v1alpha1
+kind: VirtualMachineExport
+metadata:
+    name: export-vm1
+spec:
+    source:
+        apiGroup: kubevirt.io
+        kind: VirtualMachine
+        name: vm1
+```
+
+### Export a VirtualMachine
+
+``` yaml
+apiVersion: export.kubevirt.io/v1alpha1
+kind: VirtualMachineExport
+metadata:
+    name: export-snap1
+spec:
+    source:
+        apiGroup: snapshot.kubevirt.io
+        kind: VirtualMachineSnapshot
+        name: snap1
+```
+
+### VirtualMachineExport Status
+
+``` yaml
+apiVersion: export.kubevirt.io/v1alpha1
+kind: VirtualMachineExport
+metadata:
+    namespace: ns1
+    name: export-vm1
+spec:
+    ...
+status:
+    phase: Ready # Pending|Ready|Terminated
+    token: 8fmf94kf
+    url: https://virt-import.kubevirt.my-cluster/ns1/export-vm1/download?token=8fmf94kf
+    cert: <base64 encoded cert>
+    clusterURL: https://virt-import.kubevirt/ns1/export-vm1/download?token=8fmf94kf
+    clusterCert: <base64 encoded cert>
+    conditions:
+    - type: Ready
+      status: True
+      ...
+      reason: Ready to export Virtual Machine
+```
+
+* Cluster must have Ingress/Route support for `url` and `cert` to be set
+
+## virtctl export vm|vmsnapshot
+
+virtctl will be extended to include an `export` command that does everything in the [User Flow](#user-flow) section with the addition of deleting the VirtualMachineExport when complete.
+
+```
+virtctl export vm vm1 vm1.tar.gz
+```
+
+```
+virtctl export vmsnapshot snap1 snap1.tar.gz
+```
+
+## virtctl import vm
+
+virtctl will be extended to include an `import` command that will create a new VirtualMachine based on the the contents of Virtual Machine Archive.
+
+```
+virtctl import vm import1 vm1.tar.gz
+```
+
+### Import Implementation
+
+`virtctl import` will do the following on behalf of the user:
+
+1. Validate target Virtual Machine name does not exist
+2. Validate all Persistent Volume Claim (PVC) Storage Classes exist, if not require `--storage-class-overide` args for each
+3. Generate DataVolumeTemplate with unique name for each PVC with [upload](https://github.com/kubevirt/containerized-data-importer-api/blob/main/pkg/apis/core/v1beta1/types.go#L139-L140) source
+4. Create new VirtualMachine resource with `spec.dataVolumeTemplates` populated from DataVolumeTemplates above
+5. Upload volume data from archive to target DataVolumes
+
+## Scalability
+
+Network bandwith is expected to be the biggest bottleneck for both export and import.  Given the current design, this is mostly and issue for export.  If there is a network disconnect while downloading an archive, the user will have to start again from the beginning.
```

Comment:
Can we use http ranges to allow continuation of interrupted export?

Reply:
Possibly?  But we don't have seek support on the server side so would have to read up to beginning offset

Reply:
@mhenriks could the [nbdkit retry](https://libguestfs.org/nbdkit-retry-filter.1.html) filter help here? At least to mitigate the network disconnections

Reply:
I'll take a look, but I am doubtful.  This endpoint is not simply serving up a disk image.  Rather, a dynamically constructed tarball containing multiple manifests and disk images.